src/Scenario/Merger.php

```diff
@@ -0,0 +1,232 @@
+use Magento\MagentoCloud\Filesystem\FileSystemException;
+use Magento\MagentoCloud\Filesystem\SystemList;
+use Magento\MagentoCloud\Scenario\Exception\ValidationException;
+use Symfony\Component\Serializer\Encoder\XmlEncoder;
+
+/**
+ * Merge given scenarios.
+ *
+ * @codeCoverageIgnore
+ */
+class Merger
+{
+    private const ROOT_NODE = 'scenario';
+
+    public const XSI_TYPE_STRING = 'string';
+    public const XSI_TYPE_OBJECT = 'object';
+    public const XSI_TYPE_ARRAY = 'array';
+
+    /**
+     * @var File
+     */
+    private $file;
+
+    /**
+     * @var XmlEncoder
+     */
+    private $encoder;
+
+    /**
+     * @var SystemList
+     */
+    private $systemList;
+
+    /**
+     * @var Resolver
+     */
+    private $resolver;
+
+    /**
+     * @var array
+     */
+    private static $stepRequiredArgs = [
+        '@name',
+        '@type',
+    ];
+
+    /**
+     * @var array
+     */
+    private static $argumentRequiredArgs = [
+        '@name',
+        '@xsi:type'
+    ];
+
+    /**
+     * @var array
+     */
+    private static $itemRequiredArgs = [
+        '@name',
+        '@xsi:type'
+    ];
+
+    /**
+     * @param File $file
+     * @param XmlEncoder $encoder
+     * @param SystemList $systemList
+     * @param Resolver $resolver
+     */
+    public function __construct(File $file, XmlEncoder $encoder, SystemList $systemList, Resolver $resolver)
+    {
+        $this->file = $file;
+        $this->encoder = $encoder;
+        $this->systemList = $systemList;
+        $this->resolver = $resolver;
+    }
+
+    /**
+     * @param array $scenarios
+     * @return array
+     * @throws ValidationException
+     */
+    public function merge(array $scenarios): array
+    {
+        $data = [];
+
+        foreach ($scenarios as $scenario) {
+            $scenario = $this->collectScenario($scenario);
+
+            foreach ($scenario['step'] ?? [] as $step) {
+                if ($missedArgs = array_diff(self::$stepRequiredArgs, array_keys($step))) {
+                    throw new ValidationException(sprintf(
+                        'Arguments %s are missed from step',
+                        implode(', ', $missedArgs)
+                    ));
+                }
+
+                $data[$step['@name']] = array_replace_recursive(
+                    $data[$step['@name']] ?? [],
+                    $this->collectStep($step)
+                );
+            }
+        }
+
+        return $this->resolver->resolve($data);
+    }
+
+    /**
+     * @param array $step
+     * @return array
+     * @throws ValidationException
+     */
+    private function collectStep(array $step): array
+    {
+        $arguments = [];
+
+        foreach ($step['arguments'][0]['argument'] ?? [] as $argument) {
+            if ($missedArgs = array_diff(self::$argumentRequiredArgs, array_keys($argument))) {
+                throw new ValidationException(sprintf(
+                    'Arguments %s are missed from argument',
```

Comment:
there may be 1 argument. Need to rephrase or change to "Argument(s)"

Reply:
Fixed